### DISCUSSION TITLE
Are dao, db, and mailer references meant to persist forever?

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1274] [Q&A] [answered]
@ganigeorgiev Is it going to be ok for me to peel services off of the app and to put them into my own dependency injector where they might live forever?  I mean, are any of these services meant to expire after some long period of time and then for the getter to be called again?  If not, do you plan to do that?

```golang
	dao := app.Dao()
	db := app.DB()
	mail := app.NewMailClient()
```

Basically I want to get back to how I was doing things before and to not be married to the pocketbase app struct for the entire lifetime of a long running process.  Will I have any problems with the lifetime of a variable ?

**Comment by ganigeorgiev** [comment, marked as answer]
`Dao()`, `DB()` are initialized during bootstrap and they usually don't change unless for some reason `ResetBootstrapState()` is called (_this could happens inside "plugins" for example_).

`NewMailClient()`, `NewFilesystem()`, etc., always return a new instance since they depend on the app settings and they can change during runtime.

`Settings()` also can change during runtime.

The dependency injector isn't much different than passing the result of the app getters as arguments to your functions. It of course depends on what you are using for DI, but you can also try to define them as "lazy" dependency, aka. resolve on call.

**Comment by dschissler** [comment]
This is what I came up with:

```golang
func New(app core.App) *Env {
	return &Env{
		DAO:      daoWrap(app),
	}
}

type Env struct {
	DAO      Dao
}

type Dao func() *daos.Dao

func daoWrap(app core.App) Dao {
	return func() *daos.Dao {
		return app.Dao()
	}
}
```

**Reply by dschissler** [reply]
Dang.  Does Golang automatically bind the function?  Maybe I could just do it like this to prevent the extra function call.

```golang
func daoWrap(app core.App) Dao {
	return app.Dao
}
```

**Reply by ganigeorgiev** [reply]
Yes, returning directly the `app.Dao` function should work (~~although the return type of `daoWrap` will need to change to return `func() *daos.Dao` or similar~~ I just noticed that you have declared it already).

**Reply by dschissler** [reply]
All your Pocketbase belong to us.

**Comment by dschissler** [comment]
Crazy idea.  What if `core.App` implemented/embedded `core.Services` that has this stuff?  Or something like that.

```golang
type Env struct {
	Config   *config.Config
	DAO      Dao
	DB       DB
	Mailer   Mailer
	Settings Settings
}

type Dao func() *daos.Dao
type DB func() *dbx.DB
type Mailer func() mailer.Mailer
type Settings func() *settings.Settings
```

`type Dao func() *daos.Dao` seems like a useful type.  It's de-facto everywhere in Pocketbase already.

**Reply by ganigeorgiev** [reply]
Most of the time you probably wouldn't need this, because in user methods it is expected to just pass the resolving type `*daos.Dao`, not the factory/getter function, eg:

```go
func DoSomething(dao *daos.Dao, id string) {
    dao.DB().NewQuery(...).Execute()
}
```

I don't see much benefit in further wrapping the `core.App` fields inside `Env` struct.

**Reply by dschissler** [reply]
> I don't see much benefit in further wrapping the core.App fields inside Env struct.

Well for the moment it's not doing much other than complicating things.  The idea is for me to get back to having a dependency injector.

```golang
type Dao func() *daos.Dao
```

The idea about having this getter as a built-in Pocketbase type is to have less custom-but-actually-core stuff that someone like me would need to carry around.

**Reply by dschissler** [reply]
> I don't see much benefit in further wrapping the core.App fields inside Env struct.

You're probably right about that.  But perhaps the service getter types could be useful?

**Reply by ganigeorgiev** [reply]
> But perhaps the service getter types could be useful?

I'm not sure whether it is worth doing it since you don't save much - there are no arguments just one `func()` in front of the type. It will add unnecessary "noise" to the `core` package which already has a lot of exported members, making the autocomplete suggestions confusing.

**Reply by dschissler** [reply]
I'm liking this Env struct that has bound App methods brought over.  I'm able to add in some useful helpers.

This is a bit tiresome to do when saving inside of an After hook without throwing events.

```golang
dao := daos.New(env.Dao().DB())
dao.SaveRecord(m)
```

But now I can do this:

```golang
err := env.DaoClean().SaveRecord(m)
```

because of this helper method:

```golang
func (env *Env) DaoClean() *daos.Dao {
	return daos.New(env.Dao().DB())
}
```
